extension/ackextension/inmemory.go

```diff
@@ -0,0 +1,83 @@
+package ackextension
+
+import (
+	"sync"
+)
+
+type InMemoryAckExtension struct {
+	partitionMap sync.Map
+}
+
+func NewInMemoryAckExtension() AckExtension {
+	return &InMemoryAckExtension{}
+}
+
+type ackStatus struct {
+	id     uint64
+	ackMap sync.Map
+}
+
+func newAckStatus() *ackStatus {
+	id := uint64(0)
+	ret := ackStatus{id: id}
+	ret.ackMap.Store(id, false)
+	return &ret
+}
+
+func (as *ackStatus) nextAck() uint64 {
+	as.id += 1
+	as.ackMap.Store(as.id, false)
+	return as.id
+}
+
+func (as *ackStatus) ack(key uint64) {
+	if _, ok := as.ackMap.Load(key); ok {
+		as.ackMap.Store(key, true)
+	}
+}
+
+func (as *ackStatus) queryAcks(ackIDs []uint64) map[uint64]bool {
+	result := map[uint64]bool{}
+	for _, val := range ackIDs {
+		if isAcked, ok := as.ackMap.Load(val); ok && isAcked.(bool) {
+			result[val] = true
+			as.ackMap.Delete(val)
```

Comment:
Is `queryAcks` not idempotent? Should it be renamed to `computeAcks`?

Reply:
queryAcks is not idempotent. I'll rename, thanks for the feedback!